File: backend/test/models/feedback_test.rb
require "test_helper"

class FeedbackTest < ActiveSupport::TestCase
  setup do
    @user = create_user
  end

  test "requires a message" do
    feedback = Feedback.new(user: @user, message: "")
    assert_not feedback.valid?
    assert_includes feedback.errors[:message], "can't be blank"
  end

  test "requires a user" do
    assert_not Feedback.new(message: "Something").valid?
  end

  test "rejects a message longer than the limit" do
    feedback = Feedback.new(user: @user, message: "x" * (Feedback::MESSAGE_LIMIT + 1))
    assert_not feedback.valid?
  end

  test "defaults to the new status" do
    feedback = Feedback.create!(user: @user, message: "The map pins are hard to tap")
    assert_equal "new", feedback.status
  end

  test "rejects an unknown status" do
    feedback = Feedback.new(user: @user, message: "Hi", status: "wontfix")
    assert_not feedback.valid?
  end

  test "keeps an element capture when a selector is present" do
    feedback = Feedback.create!(
      user: @user, message: "This button lies", element_selector: "main > button:nth-child(2)",
      element_classes: "_button_1p9dt_29 _quiet_1p9dt_44"
    )
    assert feedback.element?
    assert_equal "_button_1p9dt_29 _quiet_1p9dt_44", feedback.element_classes
  end

  test "drops element classes with no selector to keep the capture coherent" do
    feedback = Feedback.create!(user: @user, message: "General thought", element_classes: "_orphan_1p9dt_29")
    assert_not feedback.element?
    assert_nil feedback.element_classes
  end

  test "newest_first returns the most recent submission first" do
    older = Feedback.create!(user: @user, message: "First", created_at: 2.days.ago)
    newer = Feedback.create!(user: @user, message: "Second", created_at: 1.hour.ago)
    assert_equal [newer.id, older.id], Feedback.newest_first.pluck(:id)
  end

  test "triage runs new / in_progress / rejected / done, and nothing else" do
    assert_equal %w[new in_progress rejected done], Feedback::STATUSES
    assert_equal "new", Feedback.create!(user: @user, message: "Fresh").status

    %w[new in_progress rejected done].each do |status|
      assert Feedback.new(user: @user, message: "Fine", status: status).valid?, "#{status} should be a status"
    end

    # The name `rejected` replaced -- migration 20260828120000.
    assert_not Feedback.new(user: @user, message: "Nope", status: "triaged").valid?
  end

  test "with_statuses keeps only the statuses asked for" do
    fresh = Feedback.create!(user: @user, message: "Fresh")
    picked_up = Feedback.create!(user: @user, message: "Being worked on", status: "in_progress")
    rejected = Feedback.create!(user: @user, message: "Not acting on it", status: "rejected")
    Feedback.create!(user: @user, message: "Dealt with", status: "done")

    assert_equal [fresh.id, rejected.id].sort, Feedback.with_statuses(%w[new rejected]).pluck(:id).sort
    assert_equal [rejected.id], Feedback.with_statuses("rejected").pluck(:id)
    assert_equal [picked_up.id], Feedback.with_statuses("in_progress").pluck(:id)
  end

  # An untouched filter asks for no statuses, and nobody means "show me none of
  # them" by that -- so an empty or unrecognised list widens instead of emptying.
  test "with_statuses treats nothing recognisable as no narrowing at all" do
    3.times { |i| Feedback.create!(user: @user, message: "Note #{i}") }

    assert_equal 3, Feedback.with_statuses([]).count
    assert_equal 3, Feedback.with_statuses(nil).count
    assert_equal 3, Feedback.with_statuses(%w[triaged nonsense]).count
    # One good value among the noise still narrows by that one.
    assert_equal 3, Feedback.with_statuses(%w[new nonsense]).count
  end

  # --- Screenshots ------------------------------------------------------------
  #
  # Every case here attaches to an *unsaved* record and then asks whether it is
  # valid, which is the situation the rules have to work in: if they only held on
  # a persisted record, the file would already be in the bucket by the time it was
  # judged, and rejecting it would be too late to matter.

  test "accepts a screenshot" do
    feedback = Feedback.new(user: @user, message: "The map pins are hard to tap")
    feedback.screenshots.attach(io: file_fixture("screenshot.png").open, filename: "screenshot.png")

    assert feedback.valid?, feedback.errors.full_messages.to_sentence
    assert feedback.save
    assert_equal 1, feedback.reload.screenshots.count
    assert_equal "image/png", feedback.screenshots.first.content_type
  end

  test "accepts screenshots right up to the per-report limit" do
    feedback = Feedback.new(user: @user, message: "Five angles on one bug")
    Feedback::MAX_SCREENSHOTS.times do |i|
      feedback.screenshots.attach(io: file_fixture("screenshot.png").open, filename: "shot-#{i}.png")
    end

    assert feedback.valid?, feedback.errors.full_messages.to_sentence
  end

  test "rejects more screenshots than a report is allowed" do
    feedback = Feedback.new(user: @user, message: "Every screen in the app")
    (Feedback::MAX_SCREENSHOTS + 1).times do |i|
      feedback.screenshots.attach(io: file_fixture("screenshot.png").open, filename: "shot-#{i}.png")
    end

    assert_not feedback.valid?
    assert_includes feedback.errors[:screenshots], "are limited to #{Feedback::MAX_SCREENSHOTS} per report"
  end

  test "rejects a screenshot over the size limit" do
    feedback = Feedback.new(user: @user, message: "Here is the whole page")
    feedback.screenshots.attach(io: StringIO.new(oversized_png), filename: "huge.png", content_type: "image/png")

    assert_not feedback.valid?
    assert_includes feedback.errors[:screenshots], "must be 5 MB or smaller"
  end

  test "rejects an attachment that is not one of the allowed image types" do
    feedback = Feedback.new(user: @user, message: "Attaching my notes instead")
    feedback.screenshots.attach(io: file_fixture("not-an-image.txt").open, filename: "notes.txt")

    assert_not feedback.valid?
    assert_includes feedback.errors[:screenshots], "must be a PNG, JPEG, WebP or GIF image"
  end

  # Six wrong files are one wrong decision, so they get one sentence. Left as its
  # own test because the obvious implementation -- adding inside the loop -- passes
  # every other test here while returning the same line six times to the reporter.
  test "says what is wrong once, however many files are wrong in that way" do
    feedback = Feedback.new(user: @user, message: "Three copies of the same mistake")
    3.times { |i| feedback.screenshots.attach(io: file_fixture("not-an-image.txt").open, filename: "notes-#{i}.txt") }

    assert_not feedback.valid?
    assert_equal ["must be a PNG, JPEG, WebP or GIF image"], feedback.errors[:screenshots]
  end

  test "a report with no screenshots is untouched by the attachment rules" do
    feedback = Feedback.new(user: @user, message: "Just words")

    assert feedback.valid?
    assert_empty feedback.screenshots
  end

  # The point of validating the pending set: a rejected upload must never reach
  # storage, so nothing may have been written by the time validation fails.
  test "an unacceptable screenshot never reaches storage" do
    feedback = Feedback.new(user: @user, message: "Attaching my notes instead")
    feedback.screenshots.attach(io: file_fixture("not-an-image.txt").open, filename: "notes.txt")

    key = feedback.screenshots.attachments.first.blob.key
    assert_not feedback.save
    assert_not ActiveStorage::Blob.service.exist?(key), "the rejected file was uploaded before it was judged"
  end

  # The other half of the promise: a report and its pictures are kept together
  # or not at all. Active Storage's own sequencing commits the rows first and
  # uploads afterwards, which is exactly the case these guard against.

  test "the screenshot is in storage by the time the report is committed" do
    feedback = Feedback.new(user: @user, message: "The map pins are hard to tap")
    feedback.screenshots.attach(io: file_fixture("screenshot.png").open, filename: "screenshot.png")
    key = feedback.screenshots.attachments.first.blob.key

    assert feedback.save
    assert ActiveStorage::Blob.service.exist?(key), "the report was committed without its picture"
  end

  test "a screenshot the bucket refuses leaves nothing behind -- no report, no rows, no files" do
    feedback = Feedback.new(user: @user, message: "Two angles on one bug")
    feedback.screenshots.attach(io: file_fixture("screenshot.png").open, filename: "one.png")
    feedback.screenshots.attach(io: file_fixture("screenshot-two.png").open, filename: "two.png")
    keys = feedback.screenshots.attachments.map { |attachment| attachment.blob.key }

    # The first upload lands and the second is refused, so the test also proves
    # the one that landed is taken back out.
    with_storage_refusing(keys.last) do
      assert_no_difference [-> { Feedback.count }, -> { ActiveStorage::Blob.count }, -> { ActiveStorage::Attachment.count }] do
        assert_raises(IOError) { feedback.save }
      end
    end

    assert_not feedback.persisted?
    keys.each do |key|
      assert_not ActiveStorage::Blob.service.exist?(key), "#{key} was left in storage after the report was rolled back"
    end
  end

  # The cleanup is scoped to one save. A later save of the same object that
  # rolls back must not reach back and delete the pictures an earlier save kept.
  test "a later rollback does not delete screenshots an earlier save kept" do
    feedback = Feedback.new(user: @user, message: "Kept")
    feedback.screenshots.attach(io: file_fixture("screenshot.png").open, filename: "screenshot.png")
    assert feedback.save
    key = feedback.screenshots.first.blob.key

    feedback.status = "wontfix"
    assert_not feedback.save
    assert ActiveStorage::Blob.service.exist?(key), "a failed update deleted a picture the create had kept"
  end

  test "feedback goes with the user" do
    Feedback.create!(user: @user, message: "Bye")
    assert_difference -> { Feedback.count }, -1 do
      @user.destroy
    end
  end
end

File: frontend/src/mocks/feedback.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { db } from './db';
import type { AdminFeedback, Feedback } from '../api/types';

// The mock's job here is to be wrong in the same places the backend is wrong.
// A composer that can attach six files, or a 20 MB PNG, or a PDF, is a bug the
// UI can only discover if mock mode refuses them exactly as Rails will — so
// every limit in the contract gets a failing request of its own.

const DEMO = 1;
const SARAH = 2;

function signIn(userId: number | null) {
  db.currentUserId = userId;
}

function file(name: string, type: string, bytes = 8) {
  return new File([new Uint8Array(bytes)], name, { type });
}

function postMultipart(fields: Record<string, string>, files: File[]) {
  const form = new FormData();
  for (const [key, value] of Object.entries(fields)) form.set(`feedback[${key}]`, value);
  for (const f of files) form.append('feedback[screenshots][]', f);
  // No Content-Type header: fetch writes multipart's boundary itself, which is
  // the same thing api.postForm relies on.
  return fetch('/api/feedbacks', { method: 'POST', headers: { Accept: 'application/json' }, body: form });
}

function postJson(feedback: Record<string, unknown>) {
  return fetch('/api/feedbacks', {
    method: 'POST',
    headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
    body: JSON.stringify({ feedback }),
  });
}

beforeEach(() => signIn(DEMO));

describe('POST /api/feedbacks with screenshots', () => {
  it('stores what was attached and serializes it back', async () => {
    const response = await postMultipart(
      { message: 'The pin is in the sea', url: 'http://localhost:5173/trips/1/map' },
      [file('shot.png', 'image/png', 64), file('other.webp', 'image/webp', 32)],
    );

    expect(response.status).toBe(201);
    const { feedback } = (await response.json()) as { feedback: Feedback };
    expect(feedback.message).toBe('The pin is in the sea');
    expect(feedback.url).toBe('http://localhost:5173/trips/1/map');
    expect(feedback.screenshots).toHaveLength(2);
    expect(feedback.screenshots[0]).toMatchObject({
      filename: 'shot.png',
      content_type: 'image/png',
      byte_size: 64,
    });
    expect(feedback.screenshots[0].url).not.toBe('');
    expect(feedback.screenshots[1].filename).toBe('other.webp');
    // Distinct ids, the way separate blob records would be.
    expect(feedback.screenshots[0].id).not.toBe(feedback.screenshots[1].id);
  });

  it('keeps the JSON path exactly as it was, with an empty array', async () => {
    const response = await postJson({ message: 'Just words' });

    expect(response.status).toBe(201);
    const { feedback } = (await response.json()) as { feedback: Feedback };
    expect(feedback.screenshots).toEqual([]);
  });

  it('still normalises orphan element classes when the body is multipart', async () => {
    const response = await postMultipart({ message: 'A capture with no selector', element_classes: '_chip_7ilc4_44' }, [
      file('shot.png', 'image/png'),
    ]);

    const { feedback } = (await response.json()) as { feedback: Feedback };
    expect(feedback.element_selector).toBeNull();
    expect(feedback.element_classes).toBeNull();
  });

  it('still rejects a blank message when the files themselves are fine', async () => {
    const response = await postMultipart({ message: '   ' }, [file('shot.png', 'image/png')]);

    expect(response.status).toBe(422);
    expect(await response.json()).toEqual({ errors: { message: ["can't be blank"] } });
  });

  it('answers with everything wrong at once, each complaint said once', async () => {
    const pdfs = Array.from({ length: 6 }, (_, i) => file(`report-${i}.pdf`, 'application/pdf'));
    const response = await postMultipart({ message: '' }, pdfs);

    expect(response.status).toBe(422);
    expect(await response.json()).toEqual({
      errors: {
        message: ["can't be blank"],
        screenshots: ['are limited to 5 per report', 'must be a PNG, JPEG, WebP or GIF image'],
      },
    });
  });

  it('rejects a sixth file', async () => {
    const six = Array.from({ length: 6 }, (_, i) => file(`shot-${i}.png`, 'image/png'));
    const response = await postMultipart({ message: 'Six of them' }, six);

    expect(response.status).toBe(422);
    const body = (await response.json()) as { errors: { screenshots: string[] } };
    expect(body.errors.screenshots).toHaveLength(1);
    expect(body.errors.screenshots[0]).toMatch(/5/);
    expect(db.feedbacks.some((f) => f.message === 'Six of them')).toBe(false);
  });

  it('accepts exactly five', async () => {
    const five = Array.from({ length: 5 }, (_, i) => file(`shot-${i}.png`, 'image/png'));
    const response = await postMultipart({ message: 'Five of them' }, five);

    expect(response.status).toBe(201);
    const { feedback } = (await response.json()) as { feedback: Feedback };
    expect(feedback.screenshots).toHaveLength(5);
  });

  it('rejects a file over 5 MB', async () => {
    const tooBig = file('huge.png', 'image/png', 5 * 1024 * 1024 + 1);
    const response = await postMultipart({ message: 'One huge one' }, [tooBig]);

    expect(response.status).toBe(422);
    const body = (await response.json()) as { errors: { screenshots: string[] } };
    expect(body.errors.screenshots[0]).toMatch(/5 MB/);
  });

  it('rejects a content type that is not one of the four image formats', async () => {
    const response = await postMultipart({ message: 'A PDF, actually' }, [file('report.pdf', 'application/pdf')]);

    expect(response.status).toBe(422);
    const body = (await response.json()) as { errors: { screenshots: string[] } };
    expect(body.errors.screenshots[0]).toMatch(/PNG/);
  });

  it('is still 401 when signed out, files or no files', async () => {
    signIn(null);

    const response = await postMultipart({ message: 'Anonymous' }, [file('shot.png', 'image/png')]);

    expect(response.status).toBe(401);
  });
});

describe('DELETE /api/admin/feedbacks/:id', () => {
  function destroy(id: number) {
    return fetch(`/api/admin/feedbacks/${id}`, { method: 'DELETE', headers: { Accept: 'application/json' } });
  }

  it('removes a rejected note and answers 204 with no body', async () => {
    const response = await destroy(903);

    expect(response.status).toBe(204);
    expect(await response.text()).toBe('');
    expect(db.feedbacks.some((f) => f.id === 903)).toBe(false);
  });

  it('takes a done note too — the other finished status', async () => {
    expect((await destroy(904)).status).toBe(204);
    expect(db.feedbacks.some((f) => f.id === 904)).toBe(false);
  });

  it('refuses a note triage has not finished with, in the contract words', async () => {
    const response = await destroy(901);

    expect(response.status).toBe(422);
    expect(await response.json()).toEqual({ error: 'Only done or rejected feedback can be deleted' });
    // Refused means kept: the row is exactly where it was.
    expect(db.feedbacks.some((f) => f.id === 901)).toBe(true);
  });

  it('is 404 for an id that was never here', async () => {
    expect((await destroy(999)).status).toBe(404);
  });

  it('turns an ordinary signed-in user away, note untouched', async () => {
    signIn(SARAH);

    const response = await destroy(903);

    expect(response.status).toBe(403);
    expect(await response.json()).toEqual({ error: 'Admin access required' });
    expect(db.feedbacks.some((f) => f.id === 903)).toBe(true);
  });
});

describe('screenshots on the way out', () => {
  it('reaches the reporter through GET /api/feedbacks', async () => {
    signIn(SARAH);

    const body = (await (await fetch('/api/feedbacks', { headers: { Accept: 'application/json' } })).json()) as {
      feedbacks: Feedback[];
    };

    const seeded = body.feedbacks.find((f) => f.id === 901);
    expect(seeded?.screenshots.map((s) => s.id)).toEqual([9011, 9012]);
    expect(seeded?.screenshots[0].url).toMatch(/^data:image\/png;base64,/);
    // A row that never had any is an empty array, never undefined.
    expect(body.feedbacks.find((f) => f.id === 903)?.screenshots).toHaveLength(1);
  });

  it('reaches the admin table through GET /api/admin/feedbacks', async () => {
    const body = (await (await fetch('/api/admin/feedbacks', { headers: { Accept: 'application/json' } })).json()) as {
      feedbacks: AdminFeedback[];
    };

    expect(body.feedbacks.find((f) => f.id === 901)?.screenshots).toHaveLength(2);
    expect(body.feedbacks.find((f) => f.id === 902)?.screenshots).toEqual([]);
  });
});
